src/proxy/inbound.rs

```diff
@@ -261,38 +297,82 @@ impl Inbound {
         }
 
         let start = Instant::now();
-        let hbone_addr = req
+
+        // Extract the host or IP from the authority pseudo-header of the URI
+        let hbone_addr: HboneAddress = req
             .uri()
-            .to_string()
-            .as_str()
-            .parse::<SocketAddr>()
-            .map_err(|_| {
-                InboundError(
-                    Error::ConnectAddress(req.uri().to_string()),
-                    StatusCode::BAD_REQUEST,
-                )
-            })?;
+            .try_into()
+            .map_err(|e: Error| InboundError(e, StatusCode::BAD_REQUEST))?;
 
+        // Get the destination workload information of the destination pods (wds) workload (not destination ztunnel)
         let destination_workload = pi
             .local_workload_information
             .get_workload()
             .await
             // At this point we already fetched the local workload for TLS, so it should be infallible.
             .map_err(InboundError::build(StatusCode::SERVICE_UNAVAILABLE))?;
 
-        // Check the request is allowed
-        Self::validate_destination(&pi.cfg, &pi.state, &conn, &destination_workload, hbone_addr)
+        // Check the request is allowed by verifying the destination
+        Self::validate_destination(
+            &pi.cfg,
+            &pi.state,
+            &conn,
+            &destination_workload,
+            hbone_addr.clone(),
+        )
+        .await
+        .map_err(InboundError::build(StatusCode::BAD_REQUEST))?;
+
+        // Set to the service IP address if the hbone_addr is a hostname
+        // Used for the proxy protocol header. The Address header does not support hostname.
+        let mut upstream_protocol_addr = None;
+
+        // Validate hostname if present
+        if hbone_addr.svc_hostname().is_some() {
+            // Get the service address by hostname
+            let svc = Self::find_service_by_hostname(
+                &pi.state,
+                &destination_workload,
+                hbone_addr.svc_hostname().unwrap(),
+            )
             .await
             .map_err(InboundError::build(StatusCode::BAD_REQUEST))?;
 
+            // Validate the destination pod is a member of the selected service
+            if !svc.contains_endpoint(&destination_workload) {
+                // TODO(jaellio): Update error msg/type
+                return Err(InboundError(
+                    Error::NoHostname(destination_workload.to_string()),
+                    StatusCode::BAD_REQUEST,
+                ));
+            }
+            // Select the VIP by network match
+            let svc_address = svc
+                .vips
```

Comment:
How does this impact dualstack services that have both an IPv4 and an IPv6 VIP? Do we want to defer to v4 in some way? This looks right otherwise

Reply:
Good callout. Not sure. I'll look into this

Reply:
Updated logic prefer ipv4